lib/config_cat/client.ex

```diff
@@ -197,29 +203,26 @@ defmodule ConfigCat.Client do
   end
 
   defp entry_matching({key, setting}, variation_id) do
-    value_matching(key, setting, variation_id) ||
-      value_matching(key, Map.get(setting, Constants.rollout_rules()), variation_id) ||
-      value_matching(key, Map.get(setting, Constants.percentage_rules()), variation_id)
-  end
-
-  defp value_matching(key, value, variation_id) when is_list(value) do
-    Enum.find_value(value, nil, &value_matching(key, &1, variation_id))
-  end
-
-  defp value_matching(key, value, variation_id) do
-    if Map.get(value, Constants.variation_id(), nil) == variation_id do
-      {key, Map.get(value, Constants.value())}
+    case Setting.variation_value(setting, variation_id) do
+      nil -> nil
+      value -> {key, value}
```

Comment:
I don't see this check anywhere in the Python SDK. Should it be there as well?

Looking through the JS code, it seems like this check (or at least `isAllowedValue`) appears in many places, not just here. Should it be added elsewhere as well?

Reply:
Addressed in 5c9b4f7

Reply:
> I don't see this check anywhere in the Python SDK. Should it be there as well?

Yep, looks like Python SDK misses this check. /cc @kp-cat 

> Looking through the JS code, it seems like this check (or at least isAllowedValue) appears in many places, not just here. Should it be added elsewhere as well?

Yes, we should guarantee that when the evaluation is successful, the SDK always return a valid feature flag value (bool, string, int, double).

Reply:
Thinking it all through again, I can see the following problematic cases that can be caused by invalid flag overrides:

1. Setting type is missing from the config json or it is invalid. E.g.
    ```json
    {
      "t": 6, // unsupported type
      "v": { /*...*/ }
    }
    ```

    This is already handled here: https://github.com/configcat/elixir-sdk/blob/00289e56e29eed57e3c76afa67cc321a2494fc25/lib/config_cat/config/setting_value.ex#L24-L25

3. Setting type is valid but the return value is missing. E.g.
    ```json
    {
      "t": 0, // boolean
      "v": {
        "s": true // the wrong property is set ("b" should be set)
      }
    }
    ```

    This is handled here: https://github.com/configcat/elixir-sdk/blob/00289e56e29eed57e3c76afa67cc321a2494fc25/lib/config_cat/config/setting_value.ex#L29-L31

4. Setting type is valid, the return value is present but it is invalid. E.g.
    ```json
    {
      "t": 0, // boolean
      "v": {
        "b": "true" // this is not a boolean value!
      }
    }
    ```

   This case is not handled currently, isn't it? If my understanding is correct, "b" can be whatever garbage and will be returned to the caller.
   In such cases the evaluation will be seemingly successful, yet it returns a nonsensical value. I think we should avoid this.
   So we need to check that the type of the return value and the setting type actually match. We should do this in every case where the evaluation is successful and returns a value. (Please forget about the JS SDK's code. It wasn't a good idea to cite that because JS doesn't support complex flag overrides, so it doesn't need to handle these edge cases.)

Reply:
Addressed in 43d9d89.